File: solutions/python/perfect-numbers/perfect_numbers.py
def aliquot_sum(number):

    if number < 1:
        raise ValueError("Classification is only possible for positive integers.")

    sum = 0

    for i in range(1, number // 2 + 1):
        if number % i == 0:
            sum += i

    return sum

def classify(number):

    if number == aliquot_sum(number):
        return "perfect"
    elif number < aliquot_sum(number):
        return "abundant"
    else:
        return "deficient"

File: solutions/cpp/perfect-numbers/perfect_numbers.cpp
#include "perfect_numbers.h"

#include <stdexcept>

namespace perfect_numbers {

    auto aliquot_sum(int n) -> int
    {
        if (n < 1)
            throw std::domain_error("");

        int sum = 0;

        for (int i = 1; i <= n-1; i++)
            sum += n % i == 0 ? i : 0;

        return sum;
    }

    auto classify(int n) -> classification
    {
        if (n == aliquot_sum(n))
            return classification::perfect;
        else if (n < aliquot_sum(n))
            return classification::abundant;
        else
            return classification::deficient;
    }

}  // namespace perfect_numbers
